[1.21.9] Fix player remaining in swimming animation when head is above fluid surface

WCBBEX opened the pull request:
Fixes #2646

## Problem Description
The current logic for maintaining the swimming state in `updateSwimming()` was broadened by NeoForge to support custom fluid types. The condition `(this.isInWater() || this.isInFluidType(...))` allows a player to remain swimming as long as any part of their body is in a swimmable fluid.

This creates an issue where if the player's head is above the fluid surface (e.g., when trying to surface by holding the space bar), but their feet are still submerged, they remain stuck in the swimming animation and cannot properly transition to standing. This leads to the strange movement behavior shown in the issue's video.

## Solution
This PR introduces an additional check to the condition for maintaining the swimming state. By adding `&& this.canSwimInFluidType(this.getEyeInFluidType())`, we now ensure that the player's head (eye level) must also be within a swimmable fluid to continue swimming.

This change effectively restores the vanilla behavior of exiting the swimming state upon surfacing, while still retaining NeoForge's compatibility with custom swimmable fluids. The logic for initiating swimming remains unchanged.

### Testing
- Confirmed that the fix resolves the issue described in #2646.
- The player can now correctly surface from water and custom fluids, transitioning from swimming to standing/treading water as expected.
- Starting to swim by sprinting underwater is unaffected and works correctly.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=2662) <br/>All committers have signed the CLA.

ChampionAsh5357 commented:
Hello! Apologies for the late review. The logic seems sound, but would you mind updating your branch to target the latest to resolve any conflicts? I will do some further testing to verify after.

WCBBEX commented:
> Hello! Apologies for the late review. The logic seems sound, but would you mind updating your branch to target the latest to resolve any conflicts? I will do some further testing to verify after.

Thanks for the review! I'll fix the conflicts later today and let you know when it's updated.

WCBBEX commented:
@ChampionAsh5357 Hi, I've updated the branch and resolved the conflicts.
It looks like the Build PRs check is failing due to an unrelated issue in the upstream build script. I'll keep an eye on the main branch for a fix and will update the PR again once it's resolved. Thanks!

ChampionAsh5357 commented:
Ok, so I tested the behavior, and it doesn't perfectly replicate vanilla. Specifically, there's an inflection point and deadzone when transitioning between swimming states. When not swimming, the player needs to be within under the fluid, specifically the eyes, to start swimming. When swimming, the player needs to surface out of water for a brief moment to stop swimming. While this does fix the issue of surfacing, it also prevents the player from swimming when their eyes are above the water, which is vanilla behavior

Further testing reveals that the cause is `Entity#isUnderWater` during the set swimming as removing the method fixes the underlying issue as well without disrupting vanilla behavior. The fluid type addition changed so much that there probably is a difference in behavior somewhere for how we detect whether an entity is in a fluid; however, I was unable to backtrace the root cause. Given that `canStartSwimming` works for vanilla fluids as well, I think just removing the method is a good enough solution for now until Monica finishes their fluid rework.

WCBBEX committed:
Fix Strange player swimming behavior

WCBBEX commented:
> Ok, so I tested the behavior, and it doesn't perfectly replicate vanilla. Specifically, there's an inflection point and deadzone when transitioning between swimming states. When not swimming, the player needs to be within under the fluid, specifically the eyes, to start swimming. When swimming, the player needs to surface out of water for a brief moment to stop swimming. While this does fix the issue of surfacing, it also prevents the player from swimming when their eyes are above the water, which is vanilla behavior
> 
> Further testing reveals that the cause is `Entity#isUnderWater` during the set swimming as removing the method fixes the underlying issue as well without disrupting vanilla behavior. The fluid type addition changed so much that there probably is a difference in behavior somewhere for how we detect whether an entity is in a fluid; however, I was unable to backtrace the root cause. Given that `canStartSwimming` works for vanilla fluids as well, I think just removing the method is a good enough solution for now until Monica finishes their fluid rework.

Hi @ChampionAsh5357,
Thanks again for the detailed review and the excellent suggestion! I've updated the branch and implemented the changes you proposed by removing the isUnderWater() check. This new approach fixes the original issue without affecting the vanilla swimming behavior at the surface. It works perfectly!

extroparnishka-sudo commented:
Добавьте проверку на то что опустился ли игрок на нужный уровень в воду чтобы плыть и так же со звуком чтоб он работал если игрок может плыть
